### ISSUE TITLE
Commit 6c5bdda broke building for arm64e on Linux/iOS

### ISSUE THREAD

**Issue by L1ghtmann** [theos/theos#613]
(Thanks for reporting an issue! Please make sure you click the link above to view the issue guidelines, then fill out the blanks below.)

What are the steps to reproduce this issue?
-------------------------------------------
1. Setup Theos on an iOS device 
2. Specify the arm64e arch in a project's Makefile
3. Attempt to make

What happens?
-------------
A seg fault: 
`bash: segmentation fault 11` (Unc0ver) 
`zsh: segmentation fault  make` (Taurine)

What were you expecting to happen?
----------------------------------
The project to compile w/o error.

Any logs, error output, etc?
----------------------------
None other than the two lines above. 

Any other comments?
-------------------
I encountered this issue with both Coolstar's older clang, llvm, ld64, and odcctools (what's currently installed by the Theos Dependencies package) and with the up-to-date versions on Bingner/Elucubratus and Procursus (on their respective jailbreaks, of course).

`make troubleshoot` and `make messages=yes` also produce a seg fault, so I can't provide their outputs. 

I tried compiling multiple projects, including one made fresh with the NIC, and they all output a seg fault. 

Out of curiosity, I decided to try and manually compile a test tweak on-device (fresh out of the NIC) and didn't encounter a seg fault, so I'm pretty confident that the issue is on Theos' end. 

I don't currently have a jailbroken arm64e device to test if this occurs there as well. 

What versions of software are you using?
----------------------------------------
**Operating System:** 
iOS 14.3 (tested on both Unc0ver and Taurine on an iP7)

**Toolchain Version:**
N/A

**SDK Version:**
N/A

**Comment by UsrLightmann** [comment]
Ran into a similar issue on my Linux system just now and decided to roll back commits till I found the culprit. 

What I found was that commit [b1a7805](https://github.com/theos/theos/commit/b1a780596dcb7f7c652e0ceb653baf82c6ea869f) was the last commit where things worked as expected and commit [6c5bdda](https://github.com/theos/theos/commit/6c5bdda47624fa10ab3270578d39eb3c39b899a5) is what broke things.

**Comment by kirb** [comment]
Well that’s weird. Thanks for finding the culprit.

What happens if you do `clang --version | head -1 | perl -pe 's/^.*\(clang-(.*)\)$/$1/'` in a shell? You should get the clang version. For instance on my Mac I get `1205.0.22.9`. I have a feeling the regex may not do the job for non-Apple clang builds, and vercmp seems to infinite loop and crash make when it encounters something other than digits/periods in the string (or the string isn’t a number at all). (cc: @sbingner, in case you have ideas on that one)

**Comment by UsrLightmann** [comment]
@kirb that seems to be the issue. On both my Linux system and iP7, the aforementioned command outputs: `clang version 10.0.0 (git@git.elucubratus.com:elucubratus/packages/llvm-project.git e9aca56da1cd141e5efad838ecfe4586c53cdefa)`

**Comment by kirb** [comment]
Ah, there we go. So it definitely doesn’t match the regex. Switching it to use the upstream version (12.0.5) rather than the Apple version (1205.0.22.9) would be best.

This should print `10.0.0` for you, could you run this to confirm for me?

```
clang --version | head -1 | perl -pe 's/^.* clang version (\d+(\.\d+)+).*$/$1/'
```

**Comment by kabiroberai** [comment]
Why don't we instead utilize the fact that clang/swiftc changes the deployment target if you're targeting the arm64e v2 ABI? For example, the following
```
vtool -arch arm64e -show-build .theos/_/Library/MobileSubstrate/DynamicLibraries/MyTweak.dylib | grep -o "minos .*" | cut -d" " -f2
```
Produces `14.0` if you're building with the newer ABI. We can then compare this to the desired deployment target passed to Theos. (This should be achievable with `otool` as well; I've used `vtool` here for simplicity.)

**Comment by kabiroberai** [comment]
This works with otool (which has broader toolchain support than vtool), though I'm not entirely sure how robust it is:
```
otool -arch arm64e -l .theos/_/Library/MobileSubstrate/DynamicLibraries/MyTweak.dylib | grep "^\s*minos" | rev | cut -d" " -f1 | rev
```

**Comment by UsrLightmann** [comment]
@kirb the command you provided _does_ produce the intended output: `10.0.0` (tested on both Linux and iOS).

**Comment by UsrLightmann** [comment]
@kabiroberai the otool command you provided ran into a couple of issues:

- Linux produces the following error: `/home/lightmann/theos/toolchain/linux/iphone/bin/otool: error while loading shared libraries: libc++abi.so.1: cannot open shared object file: No such file or directory` 
- iOS produces the following error: `-sh: rev: command not found` 

Both can be fixed, of course, but the command won't work with the current state of things.

**Comment by Diatrus** [comment]
@UsrLightmann passing by: rev can be found in the `text-utils` package :)

**Comment by Diatrus** [comment]
Secondly, @kabiroberai `otool -arch arm64e -l /Library/Frameworks/Alderis.framework/Alderis-ios14 | grep -Eo '^\s*minos *[0-9]+' | grep -o '[0-9]*`, this may be a better solution as to not require vtool, and not add any new command dependencies.

**Comment by UsrLightmann** [comment]
@Diatrus Oh, cool. Don't think either are available on Elucubratus currently though, so relying on them may cause issues. 

The otool scenario in particular, since libc++-dev isn't preinstalled on Linux/provided by the toolchain and rev (via text-cmds) isn't available/preinstalled on iOS. 

The vtool scenario could work though if vtool is made available on Elucubratus through its version of odcctools (it may be available elsewhere currently and I'm just unaware).

cc: @sbingner

**Comment by Diatrus** [comment]
> @Diatrus Oh, cool. Don't think either are available on Elucubratus currently though, so relying on them may cause issues. 
> 
> The otool scenario in particular, since libc++-dev isn't preinstalled on Linux/provided by the toolchain and rev (via text-cmds) isn't available/preinstalled on iOS. 
> 
> The vtool scenario could work though if vtool is made available on Elucubratus through its version of odcctools (it may be available elsewhere currently and I'm just unaware).
> 
> cc: @sbingner  

txt-cmds is available on Procursus as a separate installation.

**Comment by UsrLightmann** [comment]
> > @Diatrus Oh, cool. Don't think either are available on Elucubratus currently though, so relying on them may cause issues. 
> > 
> > The otool scenario in particular, since libc++-dev isn't preinstalled on Linux/provided by the toolchain and rev (via text-cmds) isn't available/preinstalled on iOS. 
> > 
> > The vtool scenario could work though if vtool is made available on Elucubratus through its version of odcctools (it may be available elsewhere currently and I'm just unaware).
> > 
> > cc: @sbingner  
> 
> text-cmds is available on Procursus as a separate installation.

Yeah, I got that. I'm just saying that if the goal is to minimize dependencies, that will require two more to be added (libc++-dev for Linux and text-cmds for iOS).

**Comment by kabiroberai** [comment]
@UsrLightmann rev isn't necessary as @Diatrus pointed out, so that shouldn't be an issue. As for libc++-dev, I'm surprised that otool on Linux seems to require that. Something tells me that it should be present in the toolchain; could you try adding `<toolchain>/lib` to your LD load path before invoking otool?

**Comment by UsrLightmann** [comment]
@kabiroberai Oh, right. 

And, apologies, I'm an idiot. Libc++-dev is included with the toolchain and the issue was as you said: it wasn't being loaded. Setting LD_LIBRARY_PATH accordingly resolved the issue.

**Comment by kirb** [comment]
Fixed with #663
